From 7a25c30b15dbbb3db027d6e2fdb63d42caa6ff21 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Thu, 26 Jun 2025 09:22:55 +0200
Subject: [PATCH 973/985] evl/net: udp: return EINPROGRESS unconditionally if
 no cached route

In absence of cached route on transmit, we used to return EINPROGRESS
only if MSG_DONTWAIT was given in the request. This made it too easy
to misconfigure the transmit path - e.g. by lacking a preliminary peer
solicitation - without noticing.

Remove the condition on MSG_DONTWAIT, always return EINPROGRESS when a
packet needs to be offloaded to in-band on sendmsg() due to a route
cache miss.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 kernel/evl/net/ipv4/udp.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/kernel/evl/net/ipv4/udp.c b/kernel/evl/net/ipv4/udp.c
index c795a2ab2..a78e86ba1 100644
--- a/kernel/evl/net/ipv4/udp.c
+++ b/kernel/evl/net/ipv4/udp.c
@@ -400,13 +400,12 @@ static ssize_t send_udp(struct evl_socket *esk,
 		 * in-band stage for routing and/or MAC address
 		 * resolution. So the datagram is indeed in-flight,
 		 * but we cannot guarantee a bounded delay before it
-		 * is written to the wire. In such a case, provided
-		 * the caller asked for non-blocking I/O, return
+		 * is written to the wire. In such a case, return
 		 * -EINPROGRESS. This is a way for the caller to
 		 * detect a missing peer solicitation before the
 		 * latter is sent oob data.
 		 */
-		return unlikely(msg_flags & MSG_DONTWAIT) ? -EINPROGRESS : 0;
+		return -EINPROGRESS;
 	}
 
 	/* Ok, we have an oob path for that datagram. */
-- 
2.30.2

